spr/src/commit_message.rs

```diff
@@ -0,0 +1,426 @@
+// Write is needed by Command::Stdin.write_all()
+use std::io::Write;
+use std::process::{Command, Stdio};
+
+use crate::{
+    error::{Error, Result},
+};
+
+// Notice: use BTreeMap to make it easier to iterate trailer keys in order.
+pub type TrailerMap = std::collections::BTreeMap<String, Vec<String>>;
+
+#[derive(Debug, PartialEq)]
+pub struct CommitMessage {
+
+    /// Subject of the message (i.e. very first line)
+    pub subject:  String,
+
+    /// Body of the message, *EXCLUDING* the contents of the trailers
+    /// section. Empty string if no body.
+    pub body:     String,
+
+    /// Map of trailer keys to trailer values (e.g, "key: value...").
+    pub trailers: TrailerMap,
+}
+
+impl CommitMessage {
+
+    pub fn render(&self) -> String {
+        let mut ret: String = "".to_string();
+
+        ret.push_str(&format!("{}\n", &self.subject));
+
+        if self.body.len() > 0 {
+           ret.push_str(&format!("\n{}\n", &self.body));
+        }
+
+        if self.trailers.len() > 0 {
+            ret.push_str("\n");
+            for (k, vec) in self.trailers.iter() {
+                for v in vec.iter() {
+                    ret.push_str(&format!("{k}: {v}\n"));
+                }
+            }
+        }
+
+        ret
+    }
+}
+
+/// Parse the contents of a git commit message into a CommitMessage instance.
+pub fn parse_commit_message(
+    orig_msg: &str,
+) -> CommitMessage {
+
+    // Get rid of trailing empty/blank lines and replace all CRLFs with
+    // just LFs upfront to simplify parsing logic.
+    let msg: &str = &orig_msg
+        .trim_end()
+        .replace("\r\n", "\n");
+
+    // Parse trailers using the 'git interpret-trailers --parse` command
+    // into a trailer map.
+    let trailers = parse_trailers(msg);
+
+    // Use 1st line as the message subject and the rest as the first version
+    // of the body. The trailers paragraph, if present, will be later removed
+    // from the body.
+    let v: Vec<&str> = msg.splitn(2, "\n").collect();
+    let subject: String = v[0].to_string();
+
+    let mut body: String = "".to_string();
+    if v.len() > 1 {
+        body = v[1].to_string();
+    }
+
+    // Add back the \n to the beginning of the body so that we can look for
+    // "\n\n" when searching for the trailers paragraph.
+    body.insert(0, '\n');
```

Comment:
Please feel free to ignore - existing is very go-like so reads fine to me :)

I certainly don't have a good handle on rust idioms, but trying to eliminate the `mut`, perhaps this?
```
let body = if v.len() > 1 {
   '\n'.to_string() + v[1].to_string()
} else {
  "\n"
}

Reply:
spent a few minutes fighting the compiler with &str vs string vs. ()... 🙄 
will pass on this one, too (I have very limited b/w at this point)